### DISCUSSION TITLE
Next 13: Help implementing Server-Side components

### DISCUSSION THREAD

**Discussion by VikashLoomba** [vercel/next.js#43583] [Help]
I'm using Next 13, and many of my components feel like they should be server components, but I'm hung up on an issue.

For my auth flow, I'm making a 3rd party API request that gives me and authorization/session token that I need for all subsequent requests. The only way I can think of storing this temporary code is on the client in a Context. Obviously, this limits my ability to make API requests only on client components, rather than server components.

Is there a way I could attach this code to every outgoing request on the server? Or somehow store it in memory on the server (guessing not possible)? Perhaps there is an implementation pattern that I haven't thought of yet.

Anyone have any ideas?

**Comment by icyJoseph** [comment]
When the is auth'd, could you store a cookie on the client (with path /), so that it is present in every subsequent request to the server, and accessible via `cookies`, https://beta.nextjs.org/docs/api-reference/cookies.

**Reply by hjaber** [reply]
> When the is auth'd, could you store a cookie on the client (with path /), so that it is present in every subsequent request to the server, and accessible via `cookies`, https://beta.nextjs.org/docs/api-reference/cookies.

Sessions stored in cookies can be out of sync from the server session. Since server components can only-read cookies, this makes my auth implementation difficult as cookies/sessions expire. 

I'm curious if anyone has a good auth flow with next 13 app.

**Reply by VikashLoomba** [reply]
I ended up using a combination of setting cookies on the client side with react-cookie and utilizing middleware.ts to check if the auth token is still valid. It definitely feels hack-y, but I'll be on the lookout for better solutions.

**Reply by smo043** [reply]
@VikashLoomba - would it be possible to share the snippet? Thanks

**Reply by hjaber** [reply]
> I ended up using a combination of setting cookies on the client side with react-cookie and utilizing middleware.ts to check if the auth token is still valid. It definitely feels hack-y, but I'll be on the lookout for better solutions.

I did something very similar and it feels hacky too: client side listener for auth changes, middleware matcher to refresh auth cookie, and server components to cache session.

Using middleware and `use client` components to authenticate feels like we are missing the benefits of server components/caching.

I'm excited for more options beyond `router.refresh()` for data mutation so we can sync sessions.

**Reply by smo043** [reply]
@hjaber - could you please share a snippet? We do have a similar scenario.

**Reply by hjaber** [reply]
> @hjaber - could you please share a snippet? We do have a similar scenario.

I used the following guide for general principles:

https://supabase.com/docs/guides/auth/auth-helpers/nextjs-server-components
